### ISSUE TITLE
moz_places favicon_id not updating

### ISSUE THREAD

**Issue by foohyfooh** [crazy-max/firefox-history-merger#49]
#### Steps to reproduce this issue

1. Placed sqlite files in same directory as firefox-history-merger binary
2. firefox-history-merger merge places.sqlite places_to_merge.sqlite
3. firefox-history-merger repair-favicons places.sqlite favicons.sqlite
3. Copy new sqlite files back into Firefox profile folder

#### Expected behaviour

When a new moz_place entry is generated in favicons.sqlite, if the moz_places entry in places.sqlite has null in favicon_id field, then the favicon_id should be updated.

#### Actual behaviour

While the moz_place entry is generated in favicons.sqlite, the moz_places favicon_id field remains null so when I open history, no favicon shows up.

### Configuration

* Firefox version / arch : 80.0.1 64bit
* Operating system : Windows 10 64bit
* firefox-history-merger version (ex. 2.0.0) : 2.10.0

#### Log Output
Wed, 23 Sep 2020 16:22:40 -04 INF source/internal/app/repair-favicons.go:205 > Favicon repaired favicon_id=64532 left=21 places_id=149344 url=http://www.commitstrip.com/en/2019/03/15/css-css-everywhere/

#### Entry From moz_places
149344 | http://www.commitstrip.com/en/2019/03/15/css-css-everywhere/ | CSS, CSS everywhere | moc.pirtstimmoc.www. | 1 | 0 | 0 | 20 | 1552846539063662 | k0mi3QWolYnD | 0 | 125511498518640 | The blog relating the daily life of web agency developers | http://www.commitstrip.com/wp-content/uploads/2019/03/HeadlineImageTemplate-Conflit-5.jpg | 0

#### Entry from moz_icons
64532 | https://www.commitstrip.com/wp-content/themes/krds_blog/favicon.png?v=5 | 0 | 16 | 0 | 0 | 0 |  BLOB

**Comment by SophieFitz** [comment]
Hi @foohyfooh :) 

I know from writing my own program for combining Firefox DBs that the favicons in moz_icons won't display without a valid "fixed_icon_url_hash".

In essence, crazy-max's repair-favicons.go function will never work as it misses out this step and plugs in a default of 0 for said column. I've tested, any icon without the correct hash is invalid. I've also looked over the Firefox sourcecode, I have no idea how to replicate the hashing process. 

After I've released my Firefox History Combiner, I plan on exploring making a Firefox addon that can asynchronously (i.e. in the background while you're browsing) recreate all missing favicons. Not sure if this is even possible, and it will be slow. I also have no idea when I'll get round to this haha. 

I just wanted to let you know that recreating the icons will never work, not without a valid hash.

Also, I'm not sure I'm understanding what you're saying about favicon_id. After Firefox 55.0 the favicon_id column from moz_places (and the moz_favicons table) became redundant and was replaced entirely by the tables in favicons.sqlite. So those null values will be null forever as they're never used anyway. So in this case, the favicon_id being null is not the reason the icons aren't displaying. It's actually because of the missing hash.

Hope this made sense!
Feel free to message me, joshbrenner@protonmail.com

**Comment by mzso** [comment]
[**@JoshCode94**](https://github.com/JoshCode94) commented on [2020. szept. 25. 16:52 CEST](https://github.com/crazy-max/firefox-history-merger/issues/49#issuecomment-698975388 "2020-09-25T14:52:16Z - Replied by Github Reply Comments"):
> Hi [@foohyfooh](https://github.com/foohyfooh) :)
> 
> I know from writing my own program for combining Firefox DBs that the favicons in moz\_icons won't display without a valid "fixed\_icon\_url\_hash".
> 
> In essence, crazy-max's repair-favicons.go function will never work as it misses out this step and plugs in a default of 0 for said column. I've tested, any icon without the correct hash is invalid. I've also looked over the Firefox sourcecode, I have no idea how to replicate the hashing process.
> 
> After I've released my Firefox History Combiner, I plan on exploring making a Firefox addon that can asynchronously (i.e. in the background while you're browsing) recreate all missing favicons. Not sure if this is even possible, and it will be slow. I also have no idea when I'll get round to this haha.
> 
> I just wanted to let you know that recreating the icons will never work, not without a valid hash.
> 
> Also, I'm not sure I'm understanding what you're saying about favicon\_id. After Firefox 55.0 the favicon\_id column from moz\_places (and the moz\_favicons table) became redundant and was replaced entirely by the tables in favicons.sqlite. So those null values will be null forever as they're never used anyway. So in this case, the favicon\_id being null is not the reason the icons aren't displaying. It's actually because of the missing hash.
> 
> Hope this made sense!  
> Feel free to message me, [joshbrenner@protonmail.com](mailto:joshbrenner@protonmail.com)

Hi!
So is this why I don't have favicons? Did you find a solution in the time since?

It's weird, because from the website some items have favicon, while most don't, such as:
![kép](https://user-images.githubusercontent.com/4479090/159125505-c0bf539b-60b7-4fa7-9e29-49dc3537d4e9.png)
I expected all items form a domain having the same favicon.

**Comment by SophieFitz** [comment]
> 
> Hi! So is this why I don't have favicons? Did you find a solution in the time since?

Hi @mzso :)
I would imagine so yes. I have found a solution and I have since uploaded my project to Github but it's not finished. I have had other priorities in recent months but I would like to finish the program up soon! Here's the link: [Firefox History Combiner](https://github.com/JoshCode94/Firefox-History-Combiner)

> It's weird, because from the website some items have favicon, while most don't, such as: ![kép](https://user-images.githubusercontent.com/4479090/159125505-c0bf539b-60b7-4fa7-9e29-49dc3537d4e9.png) I expected all items form a domain having the same favicon.

As far some icons displaying and others not, sometimes one domain will have different icons depending on the url.
Cheers,
Josh

**Comment by mzso** [comment]
@JoshCode94 
Thanks for the reply. I looked there, but I only saw a remove duplicate entries program among the releases.
I guess I should have looked at the python script? (For which I'm not sure I have the competence to use.)

I since let go of it (wasted too much time already), and just used the older favicons.sqlite, which only resulted in missing favicons for this month's history.
Though it would still be nice to be able to fetch favicons for those items that don't have one. (I don't have high hopes)

**Comment by SophieFitz** [comment]
> @JoshCode94 Thanks for the reply. I looked there, but I only saw a remove duplicate entries program among the releases. I guess I should have looked at the python script? (For which I'm not sure I have the competence to use.)

Yeah the program's not finished, only the remove duplicates script is released right now. I just wanted to direct you to my project for future reference. I don't when it'll be done but it will be! Keep checking back.

> I since let go of it (wasted too much time already), and just used the older favicons.sqlite, which only resulted in missing favicons for this month's history.

That's fair.
> Though it would still be nice to be able to fetch favicons for those items that don't have one. (I don't have high hopes)

It is possible, but it's complicated and slow and I have not tackled it yet (it is on the non-priority part of my todo list).
